server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java

```diff
@@ -324,6 +292,74 @@ public void testLimitOnExpandedFieldsButIgnoreUnmappedFields() throws Exception
         client().prepareSearch("ignoreunmappedfields").setQuery(qb).get();
     }
 
+    public void testLimitOnExpandedFields() throws Exception {
+        XContentBuilder builder = jsonBuilder();
+        builder.startObject();
+        {
+            builder.startObject("_doc");
+            {
+                builder.startObject("properties");
+                {
+                    for (int i = 0; i < CLUSTER_MAX_CLAUSE_COUNT; i++) {
+                        builder.startObject("field_A" + i).field("type", "text").endObject();
+                        builder.startObject("field_B" + i).field("type", "text").endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+        }
+
+        assertAcked(prepareCreate("testindex")
+                .setSettings(Settings.builder().put(MapperService.INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING.getKey(),
+                        CLUSTER_MAX_CLAUSE_COUNT + 100))
+                .setMapping(builder));
+
+        client().prepareIndex("testindex").setId("1").setSource("field_A0", "foo bar baz").get();
+        refresh();
+
+        // single field shouldn't trigger the limit
+        doAssertOneHitForQueryString("field_A0:foo");
+        // expanding to the limit should work
+        doAssertOneHitForQueryString("field_A\\*:foo");
+        // expanding two blocks to the limit still works
+        doAssertOneHitForQueryString("field_A\\*:foo field_B\\*:bar");
```

Comment:
Just out of curiosity: this test shows what to me is a bit counter-intuitive. When we check breaching the limit once per expansion but not in its sum we can easily hit more than `indices.query.bool.max_clause_count` with a query.